conductor_api/src/conductor/base.rs

```diff
@@ -653,6 +605,77 @@ impl Conductor {
             })
     }
 
+    pub fn build_conductor_api(
+        &mut self,
+        instance_id: String,
+        config: &Configuration,
+    ) -> Result<IoHandler, HolochainError> {
+        let instance_config = config.instance_by_id(&instance_id)?;
+        let agent_id = instance_config.agent.clone();
+        let agent_config = config.agent_by_id(&agent_id)?;
+        let mut api_builder = ConductorApiBuilder::new();
+        // Signing callback:
+        if let Some(true) = agent_config.holo_remote_key {
+            // !!!!!!!!!!!!!!!!!!!!!!!
+            // Holo closed-alpha hack:
+            // !!!!!!!!!!!!!!!!!!!!!!!
+            api_builder = api_builder.with_outsource_signing_callback(
+                self.agent_config_to_id(&agent_config)?,
+                self.config
+                    .signing_service_uri
+                    .clone()
+                    .expect("holo_remote_key needs signing_service_uri set"),
+            );
+        } else {
+            api_builder = api_builder.with_agent_signature_callback(
+                self.get_keybundle_for_agent(&instance_config.agent)?,
+            );
+
+            let keystore = self
+                .get_keystore_for_agent(&instance_config.agent)
+                .map_err(|err| format!("{}", err))?;
+            api_builder = api_builder.with_agent_keystore_functions(keystore);
+        }
+
+        // Bridges:
+        let id = instance_config.id.clone();
+        for bridge in config.bridge_dependencies(id.clone()) {
+            assert_eq!(bridge.caller_id, id.clone());
+            let callee_config = config
+                .instance_by_id(&bridge.callee_id)
+                .expect("config.check_consistency()? jumps out if config is broken");
+            let callee_instance = self.instances.get(&bridge.callee_id).expect(
+                r#"
+                    We have to create instances ordered by bridge dependencies such that we
+                    can expect the callee to be present here because we need it to create
+                    the bridge API"#,
+            );
+
+            api_builder =
+                api_builder.with_named_instance(bridge.handle.clone(), callee_instance.clone());
+            api_builder =
+                api_builder.with_named_instance_config(bridge.handle.clone(), callee_config);
```

Comment:
Not to be applied but you could also probably make use of the question mark operator here by folding

config.bridge_dependencies(id.clone()).iter().fold(OK(api_builder),|bridge,current_result|{
current_result?;
config.instance_by_id(&bridge.callee_id)?;
self.instances.get(&bridge.callee_id)?;
api_builder=api_builder.with_named_instance(bridge.handle.clone(), callee_instance.clone());
api_builder =api_builder.with_named_instance_config(bridge.handle.clone(), callee_config);
OK(api_builder)
})?;

Reply:
Just to reduce the number of expects here

Reply:
The expect above is one of those few examples were I really think an expect is a more appropriate choice than returning an error. The reason is that it really is an invariant of the code that the callee instance should be live already. Instances are created ordered by their bridge dependencies. If this invariant should be violated through some future changes, I want an `expect` to notify the developer. It is not a runtime error that could be recovered from by a developer mistake. That is exactly the semantic attached to explicit panics. Maybe an `assert!` would make this even clearer?

Reply:
Assert would be clearer I think because returning a result but doing expect makes it look like we are going around the type system.

I guess it is also a question of "what do we do with the result of this whole process?" cause if we expect it then we could also return a result and then expect it from the outside? I think if it is developer notification then  yes maybe an assert would be better